flink-runtime/src/main/java/org/apache/flink/runtime/rpc/RpcEndpoint.java

```diff
@@ -172,6 +173,13 @@ public void runAsync(Runnable runnable) {
 		return ((MainThreadExecutor) self).callAsync(callable, timeout);
 	}
 
+	/**
+	 * Returns the class of the self gateway type.
+	 *
+	 * @return Class of the self gateway type
+	 */
+	public abstract Class<C> getSelfGatewayType();
```

Comment:
Yes we can do this. I thought about this and then went for the solution which saves some compute cycles. I can change it to the reflection alternative if it's pre-mature optimization.